### PR TITLE
Add a QQ provider

### PR DESCRIPTION
**Pull request by zhengchun**
#138 
Test successful

### CODE REVIEW

File: src/AspNet.Security.OAuth.QQ/QQAuthenticationHandler.cs [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+            : base(client)
+        {
+        }
+
+        protected override async Task<AuthenticationTicket> CreateTicketAsync(ClaimsIdentity identity, AuthenticationProperties properties, OAuthTokenResponse tokens)
+        {
+            var openid = await GetUserOpenIdAsync(tokens);
+            var queryString = new Dictionary<string, string>()
+            {
+                {"oauth_consumer_key",Options.ClientId },
```
Comment (kevinchalet) [inline src/AspNet.Security.OAuth.QQ/QQAuthenticationHandler.cs:36]: `oauth_consumer_key` is an OAuth1 thing. Are you sure QQ supports OAuth2?

File: src/AspNet.Security.OAuth.QQ/QQAuthenticationHandler.cs [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+            {
+                var msg = payload.Value<string>("msg");
+                throw new HttpRequestException($"An error occurred when retrieving user information. code: {ret}, msg: {msg}");
+            }
+
+            identity.AddOptionalClaim(ClaimTypes.NameIdentifier, openid, Options.ClaimsIssuer)
+                    .AddOptionalClaim(ClaimTypes.Name, QQAuthenticationHelper.GetNickname(payload), Options.ClaimsIssuer)
+                    .AddOptionalClaim(ClaimTypes.Gender, QQAuthenticationHelper.GetGender(payload), Options.ClaimsIssuer)
+                    .AddOptionalClaim("urn:qq:figureurl", QQAuthenticationHelper.GetFigureUrl(payload), Options.ClaimsIssuer)
+                    .AddOptionalClaim("urn:qq:figureurl_1", QQAuthenticationHelper.GetFigureUrl1(payload), Options.ClaimsIssuer)
```
Comment (kevinchalet) [inline src/AspNet.Security.OAuth.QQ/QQAuthenticationHandler.cs]: Not sure I like these _1/_2/qq_1/qq_2 claims. Do we really need them? What do they do?

File: src/AspNet.Security.OAuth.QQ/QQAuthenticationHandler.cs [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+        private async Task<string> GetUserOpenIdAsync(OAuthTokenResponse tokens)
+        {
+            var endpoint = QueryHelpers.AddQueryString(Options.OpenIdEndpoint, "access_token", tokens.AccessToken);
+            var response = await Backchannel.GetAsync(endpoint);
+            if (!response.IsSuccessStatusCode)
+            {
+                throw new HttpRequestException($"An error occurred while retrieving the user openid.({response.StatusCode})");
+            }
+            var body = await response.Content.ReadAsStringAsync();
+            var i = body.IndexOf("{");
```
Comment (kevinchalet) [inline src/AspNet.Security.OAuth.QQ/QQAuthenticationHandler.cs:110]: That sounds fragile. What's the purpose of this thing?

**Review by kevinchalet — CHANGES_REQUESTED** [review]

**Comment by kevinchalet** [comment]
@zhengchun thanks for your PR (and sorry for the late review).

I added a few remarks, but overall it looks good.

File: src/AspNet.Security.OAuth.QQ/QQAuthenticationHandler.cs (hunk 1 above)
Comment (zhengchun) [reply src/AspNet.Security.OAuth.QQ/QQAuthenticationHandler.cs:36]: Yes, QQ supports OAuth2 and test success.

File: src/AspNet.Security.OAuth.QQ/QQAuthenticationHandler.cs (hunk 3 above)
Comment (zhengchun) [reply src/AspNet.Security.OAuth.QQ/QQAuthenticationHandler.cs:110]: The QQ OAuth server responses openid content format like this: `callback( {“client_id”:”YOUR_APPID”,”openid”:”YOUR_OPENID”} );`

File: src/AspNet.Security.OAuth.QQ/QQAuthenticationHandler.cs (hunk 2 above)
Comment (zhengchun) [reply src/AspNet.Security.OAuth.QQ/QQAuthenticationHandler.cs]: Sorry, its my fault, I'm add new commit.

**Comment by kevinchalet** [comment]
Merged: https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers/commit/c8c923a5b6149fa7693bd3a54bbcbc988fc1a56e

Thanks for your contribution :+1:

**Pull request closed without merging**